static/js/controllers/contacts-edit.js

```diff
@@ -192,142 +193,121 @@ angular.module('inboxControllers').controller('ContactsEditCtrl',
+      _.each(schema.fields, function(fieldConf, fieldName) {
+        var customType = fieldConf.type.match(/^(db|custom):(.*)$/);
         if (customType) {
-          result = result.then(function(doc) {
-            if(!doc[f]) {
-              doc[f] = {};
-            } else if(doc[f] === 'NEW') {
-              var extra = extras[f];
-              extra.type = customType[2];
-              extra.reported_date = Date.now();
-
-              var isChild = extra.parent === 'PARENT' ||
-                            (!extra.parent && conf.parent === 'PARENT');
-              if (isChild) {
-                delete extra.parent;
-              }
-
-              return persist(extra)
-                .then(function(newlySavedDoc) {
-                  doc[f] = newlySavedDoc;
-                  if(isChild) {
-                    children.push(newlySavedDoc);
-                  }
-                  return doc;
-                });
-            } else if(original && original[f] && doc[f] === original[f]._id) {
-              doc[f] = original[f];
+          if(!doc[fieldName]) {
+            doc[fieldName] = {};
+          } else if(doc[fieldName] === 'NEW') {
+            var preparedSibling = prepare(siblings[fieldName]);
+            preparedSibling.type = customType[2];
+
+
+            var isChild = preparedSibling.parent === 'PARENT' ||
+                          (!preparedSibling.parent && fieldConf.parent === 'PARENT');
+            if (isChild) {
+              delete preparedSibling.parent;
+              // Cloning to avoid the circular reference we would make:
+              //   doc.fieldName.parent.fieldName.parent...
+              doc[fieldName] = _.clone(preparedSibling);
+              preparedSibling.parent = doc;
             } else {
-              var docId = doc[f];
-              if (typeof docId === 'object') {
-                docId = doc[f]._id;
-              }
-              return DB().get(doc[f])
-                .then(function(dbFieldValue) {
-                  doc[f] = dbFieldValue;
-                  return doc;
-                });
+              doc[fieldName] = preparedSibling;
             }
-            return doc;
-          });
+
+            preparedSiblings.push(preparedSibling);
+          } else if(original &&
+                    original[fieldName] &&
+                    original[fieldName]._id === doc[fieldName]) {
+
+            doc[fieldName] = original[fieldName];
+          } else {
+            throw new Error('Unsupported operation, talk to your friendly dev');
+            // REVIEWER:
+            //
+            // I don't think this branch is actually exercised, when:
+            //   var doc = doc[fieldName];
+            //  either:
+            //   if (typeof docId === 'object')
+            // is misleading and never exercised, or there is a bug in:
+            //   DB().get(doc[fieldName])
+            // since doc[fieldName] it could be an object not a string _id
```

Comment:
So the entire original code (without my var renaming) is this: 
```js
var docId = doc[f];
if (typeof docId === 'object') {
  docId = doc[f]._id;
}
return DB().get(doc[f])
  .then(function(dbFieldValue) {
    doc[f] = dbFieldValue;
    return doc;
  });
```

So `docId === doc[f]`, and `docId` can be an object sometimes. That would then mean `DB().get(doc[f])` was sometimes passing an object to get. More importantly, `docId` is never actually used after its set. I presume it's *supposed* to read `return DB().get(docId)`.

So to me that either means:
 - `doc[f]` can be a string id but *never* be an object, and we replace that id with an object sometimes (I'd love an example of this, because if I could get rid of it I could keep the code synchronous :P)
 - the code is never called at all

The options for `doc[f]` in that entire if statement block are:
 - undefined, in which case it becomes `{}` (why!?!?)
 - `"NEW"`, in which case we attach an "extra" / sibling we absolutely must find at that location
 - an id which is the same as the original document, in which case we just use the attached object from the original
 - anything else, in which case if it's a string we attempt to load that object and attach it, and otherwise (IMO, see above) we crash

I understand and know examples of the first two, but the last 2-3 are mysteries to me.